Downward API proposal for resources (cpu, memory) limits and requests

aveshagarwal opened the pull request:
Proposal to address #9473
This PR proposes three approaches to expose values of resource limits and requests as env vars and volumes.This proposal has details about merits and demerits of each approach, and I am looking for community feedback regarding which one (or may more than one) we would like to go with. Also would like to know if there is any other approach.

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24051)

<!-- Reviewable:end -->

aveshagarwal committed:
Downward API proposal for resources (cpu, memory) limits and requests

aveshagarwal commented:
@kubernetes/rh-cluster-infra

aveshagarwal commented:
For reference: https://github.com/kubernetes/kubernetes/issues/9473

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+## Background
+
+Currently downward APIs (via environment variables and volume plugin) only support exposing a Pod's name, namespace, annotations, labels and its IP. This document explains the need and design to extend them to expose resources (e.g. cpu, memory, storage) limits and requests.
```
We just call it 'the downward API'

bgrant0607 replied:
Please link to the main downward API doc on kubernetes.io

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+## Design
+
+This is mostly driven by the discussion in https://github.com/kubernetes/kubernetes/issues/9473.
```
This is the right place to summarize the different approaches considered in the proposal

vishh replied:
+1

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+Since envionment variables are container scoped, so there is no need to specify container name as part of the partial selectors as they are relative to container spec.
+
+#### Volume plugin
```
Partial paths are not going to work for volume plugins; suggest dropping this section.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+## Motivation
+
+By exposing information about computing resource requests & limits into pods, image/application authors can be aware of the computing environment in which the pods run, and could modify container images accordingly.
```
I would restate this a little.  How about: "It is common for software packages to configure limitations on their resource usage.  Exposing the requested and limited amounts of available resources will allow packages that do this to be configured more easily.

You can also state that docker already exposes some of this information, but we want to be runtime-agnostic in Kubernetes.

vishh replied:
Existing applications are used to fetching resources from either procfs or cgroupfs. Should we just wait for cgroup namespaces to be available instead?

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+Volumes are pod scoped, so a selector should be specified with a particular container name.
+
+Note: environment variables and volume path names could be anything and not necessarily as specified above.
```
s/could be anything.../are examples only/

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+## Output Format
+
+The output format for resources limits and requests will be same as cgroups output format.
```
Say what that format is

pmorie commented:
@aveshagarwal What's your opinion about which option to support?  What are the relative merits and drawbacks?  A proposal should examine the different options, contrast them on the merits, and pick one.  So far this is a good explanation of the options, but I think it could use more comparison/contrast/and opinion.

bgrant0607 commented:
cc @justinsb @vishh @dchen1107

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+| ---- | ------------------- |
+| CPU_LIMIT | .spec.containers[?(@.name=="container-name")].resources.limits.cpu |
+| MEMORY_LIMIT | .spec.containers[?(@.name=="container-name")].resources.limits.memory |
+| STORAGE_LIMIT | .spec.containers[?(@.name=="container-name")].resources.limits.storage |
```
We don't have storage limits and requests.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+| CPU_REQUEST | .spec.containers[?(@.name=="container-name")].resources.requests.cpu |
+| MEMORY_REQUEST | .spec.containers[?(@.name=="container-name")].resources.requests.memory |
+| STORAGE_REQUEST | .spec.containers[?(@.name=="container-name")].resources.requests.storage |
+
```
Examples of how to use these to set the Java heap size and GOMAXPROCS would be helpful.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+### API with full json path selectors
+
+Full json path selectors specify the complete path to the resources limits and requests relative to pod spec.
```
Is this identical to the dialect supported by kubectl?

aveshagarwal replied:
yes

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+| Name | Selectors |
+| ---- | ------------------- |
+| CPU_LIMIT | .spec.containers[?(@.name=="container-name")].resources.limits.cpu |
```
Existing uses (e.g. metadata.name) don't start with dot.

aveshagarwal replied:
They dont have to start with dot. I will clarify that in next update of this proposal.

vishh commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+
+#### Environment variables
+
+This table shows how selectors can be used for various requests and limits to be exposed as environment variables.
```
How would this work if we support dynamically updating resource limits in the future?
cc @bgrant0607

bgrant0607 replied:
IMO, in general, if a user updates some property that's fed into an env var via downward API, then it should cause the container to restart. If the user doesn't want the container to restart, they can use volumes instead or the API. 

But it's sort of moot right now.

vishh replied:
@bgrant0607 Is it fair to assume that resource limits/requests can change often in the future, if we have vertical autoscaling? If yes, should we even allow exposing these values via env vars?

bgrant0607 replied:
@vishh We need a way to set Java heap size and GOMAXPROCS on the command line. I'm more worried about making it possible to run common workloads now. I think we're far, far away from enabling vertical autosizing by default. Users who opt in will need to make other changes to their apps in order to adequately take advantage of it. Apps aren't built that way today.

thockin commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,262 @@
+| container_name/memory_request | resources.requests.memory |
+| container_name/storage_request | resources.requests.memory |
+
+Since envionment variables are container scoped, so container name must be specified as part of their path name. The format is that container name and path name must be separated by slash (`/`).
```
This is really ambiguous and not safe if/when we add pod-level resources

aveshagarwal replied:
@thockin thanks for your comment. Could you please elaborate on "not safe" part?

thockin replied:
if we have a pod-scoped resources block, it's not clear whether this means pod resources or "this container" resources.  Not future safe

aveshagarwal replied:
I agree.  Though my understanding is that as there are 2 field references in this proposal: one is fieldRef (already exists) which is relative to pods, and  a new containerFieldRef relative to containers,  the above scenario is supposed to be used only with containerFieldRef as explained in the proposal.

So if a pod-scoped resourced block is added in future, that should only be used with fieldRef not with containerFieldRef, something like:

```
      downwardAPI:
        items:
          - path: "pod-resource-whatever"
            fieldRef:
              fieldPath: spec.pod-resources.whatever
```
Even if someone specifies it incorrectly with containerFieldRef, that would exit with error.

thockin replied:
Ahh, if it is disambiguated that way, then OK

On Mon, Apr 18, 2016 at 11:51 AM, Avesh Agarwal <notifications@github.com>
wrote:

> In docs/design/downward_api_resources_limits_requests.md
> <https://github.com/kubernetes/kubernetes/pull/24051#discussion_r60113412>
> :
>
> > +| STORAGE_REQUEST | resources.requests.storage |
> > +
> > +Since envionment variables are container scoped, so there is no need to specify container name as part of the partial selectors as they are relative to container spec.
> > +
> > +#### Volume plugin
> > +
> > +| Path | Selectors |
> > +| -------------------- | ------------------- |
> > +| container_name/cpu_limit | resources.limits.cpu |
> > +| container_name/memory_limit | resources.limits.memory |
> > +| container_name/storage_limit | resources.limits.storage |
> > +| container_name/cpu_request | resources.requests.cpu |
> > +| container_name/memory_request | resources.requests.memory |
> > +| container_name/storage_request | resources.requests.memory |
> > +
> > +Since envionment variables are container scoped, so container name must be specified as part of their path name. The format is that container name and path name must be separated by slash (`/`).
>
> I agree. Though my understanding is that as there are 2 field references
> in this proposal: one is fieldRef (already exists) which is relative to
> pods, and a new containerFieldRef relative to containers, the above
> scenario is supposed to be used only with containerFieldRef as explained in
> the proposal.
>
> So if a pod-scoped resourced block is added in future, that should only be
> used with fieldRef not with containerFieldRef, something like:
>
>       downwardAPI:
>         items:
>           - path: "pod-resource-whatever"
>             fieldRef:
>               fieldPath: spec.pod-resources.whatever
>
> Even if someone specifies it incorrectly with containerFieldRef, that
> would exit with error.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24051/files/75618427618df97a3b5b345193f1fe3ed0ef95d7#r60113412>
>

aveshagarwal commented:
I have updated this proposal with more details and addressed all comments except following 2: pmorie's cgroup output format and bgrant's Examples of how to use these to set the Java heap size and GOMAXPROCS would be helpful. I will incorporate them soon.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,436 @@
+
+3. In this approach, users specify particular strings to retrieve resources limits and requests and do not specify any json path selectors. This approach is similar to the existing downward API implementation approach. The advantages of this approach is that it is simpler to specify that the firs two, and does not require any type of conversion between internal and versioned objects or json selec ...
+
+Before discussing a bit more about merits of each approach, here is a brief discussion about json path selectors and some implications related to their use.
```
make a subheading for this 'A note about JSONPath selectors'

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,436 @@
+
+So there is a one time conversion cost associated with the first (full path) and second (partial path) approaches, whereas the third approach (no selectors) does not require any such conversion and can directly work on internal objects. If we want to avoid conversion cost and to have implementation simplicity, my opinion is that no selector approach is relatively easiest to implement to expose li ...
+
+The current implementation ([PR 24179](https://github.com/kubernetes/kubernetes/pull/24179)) of this proposal uses the third (no selectors) approach, and is similar to the existing downward API implementation.
```
Remove this note, your current WIP code isn't a factor in the proposal :)

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,435 @@
+
+## Background
+
+Currently downward API (via environment variables and volume plugin) only support exposing a Pod's name, namespace, annotations, labels and its IP ([see details](http://kubernetes.io/docs/user-guide/downward-api/)). This document explains the need and design to extend them to expose resources (e.g. cpu, memory) limits and requests.
```
'the downward API only supports'

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,435 @@
+
+## Motivation
+
+Software applications generally configure limitations on their resource usage for their efficient operations. Exposing the requested and limited amounts of available resources inside containers will allow these applications to be configured more easily. Although docker already exposes some of this information inside containers, the downward API helps exposing this information in a runtime-agnosti ...
```
`Software applications require configuration to work optimally with the resources they're allowed to use.`

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,435 @@
+
+This is mostly driven by the discussion in [this issue](https://github.com/kubernetes/kubernetes/issues/9473). There are three approaches discussed in this document to obtain resources limits and requests to be exposed as environment variables and volumes inside containers:
+
+1. The first approach requires users to specify full json path selectors in which selectors are relative to the pod spec. The benefit of this approach is to specify pod level resources, and since containers are also part of a pod spec, it can be used to specify container level resources too.
```
s/in which selectors are/which are/

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,435 @@
+
+2. The second approach requires specifying partial json path selectors in which selectors are relative to the container spec. This approach helps in retrieving a container specific resource limits and requests, and at the same time, it is simpler to specify than full json path selectors.
+
+3. In this approach, users specify particular strings to retrieve resources limits and requests and do not specify any json path selectors. This approach is similar to the existing downward API implementation approach. The advantages of this approach is that it is simpler to specify that the firs two, and does not require any type of conversion between internal and versioned objects or json selec ...
```
The first couple sentences sound a little off.  How about:  `The third approach avoids selectors entirely and introduces a new object that has explicit fields for container name, limit type, and request type.`

pmorie commented:
@aveshagarwal can you please wrap lines at 80 chars?  Run-on lines are tough to comment on in github.

aveshagarwal commented:
@pmorie  In my browser, I dont see Run-on lines, everything is wrapped up.

pmorie commented:
@aveshagarwal the problem is i can only make one comment for those run-on-lines.

aveshagarwal commented:
@pmorie, so you mean splitting lines in a paragraph?

pmorie commented:
@aveshagarwal yeah, your text editor can probably do it for you -- for example in sublime, ctrl+\ just wraps a block of text automatically

aveshagarwal commented:
@pmorie, Wrapping at 80 chars messed up other things, fixing them, will do another update soon.

pmorie commented:
@aveshagarwal thanks

aveshagarwal commented:
Pushed an update.

bgrant0607 commented:
Please run hack/update-generated-docs.sh

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+brief discussion about json path selectors and some implications related
+to their use.
+
+#### JSONpath selectors
```
@lavalamp May be interested in this part about conversion of jsonpath field selectors.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+| Path | Selectors |
+| ---- | ------------------- |
+| cpu_limit |spec.containers[?(@.name=="container-name")].resources.limits.cpu|
+| memory_limit |spec.containers[?(@.name=="container-name")].resources.limits.cpu|
```
All the selectors say cpu. 2 of them should be memory.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+        // Required: Version of the schema the FieldPath is written in terms of.
+        // If no value is specified, it will be defaulted to the APIVersion of the
+        // enclosing object.
+        APIVersion string `json:"apiVersion"`
```
This field isn't required. It's optional, since it has a default value.

aveshagarwal replied:
Right, just put it to show how the type ObjectFieldSelector struct looks like.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+| CPU_LIMIT | resources.limits.cpu |
+| MEMORY_LIMIT |resources.limits.memory |
+| CPU_REQUEST | resources.requests.cpu |
+|MEMORY_REQUEST | resources.requests.memory |
```
inconsistent spacing

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
````diff
@@ -0,0 +1,530 @@
+```
+// Represents a single file containing information from the downward API
+type DownwardAPIVolumeFile struct {
+     // Required: Path is  the relative path name of the file to be created. Must not be absolute or contain the '..' path. Must be utf-8 encoded. The first item of the relative path must not start with '..'
````
The comment is confusing: Must not be absolute or contain the '..'

Does that mean it must or can't contain `..` if not absolute? If the latter, please s/or/nor/ and remove the redundant sentence about not starting with `..`.

aveshagarwal replied:
I agree and I will fix it in next update, thanks.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+in retrieving a container specific resource limits and requests, and at
+the same time, it is simpler to specify than full json path selectors.
+
+3. In this approach, users specify particular strings to retrieve
```
The third approach

By "particular strings", you mean "ad hoc strings"?

aveshagarwal replied:
I mean "fixed" strings.

bgrant0607 replied:
I think you mean "magic" strings. In what way is that similar to the current downward API approach? The user specifies paths, such as `metadata.name`.

aveshagarwal replied:
Yes. As per my understanding, these magic strings are treated just as strings not json path (though they "look" same as json paths) in current downward API implementation that means strings (or parts of the strings) are not processed (and at the same time not validated) against the actual json tags in the versioned kube objects. The similar approach is used in the 3rd approach in this proposal, and thats why I call them no json path selectors, again even though they "look" same as json path.  

To process strings as json paths, one would have to use go reflections (I am not sure if there is any other way) as is done in json path library and is used in kubectl. One example is that if lets say If change json tags `metadata` and `name` to something like `metapart` and `realname`, the magic string "metadata.name" would keep working in the current downward API, but will fail in kubectl as `metadata.name` is incorrect now and the correct one is `metapart.realname` . 

Please understand I am `Not` saying the current downward API is wrong (the current approach is perfectly fine), I am just saying it does not process json paths as json paths but as magic strings as is the 3rd approach, but is very different from the 1st and 2nd approaches.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+resources limits and requests and do not specify any json path
+selectors. This approach is similar to the existing downward API
+implementation approach. The advantages of this approach is that it is
+simpler to specify that the firs two, and does not require any type of
```
first

bgrant0607 commented:
I assume we're looking for a decision about which of the 3 approaches to take? That would have been useful to call out in the PR description.

aveshagarwal commented:
@bgrant0607 Right,  I will update this PR description.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+
+### API with no selectors
+
+In this approach, users specify particular strings to retrieve resources
```
This isn't what I had thought we had described during our conversation on Friday.  For example, this approach still uses field selectors.  I thought we had discussed something like:

```go
type ResourceRef struct {
  ContainerName string
  Request       ResourceType
  Limit         ResourceType
}

type ResourceType string

const (
  ResourceTypeCPU    ResourceType = "cpu"
  ResourceTypeMemory ResourceType = "memory"
)
```

@aveshagarwal

aveshagarwal replied:
@pmorie Those field selectors are not really selectors in the sense of json path selectors as they are processed differently as "fixed" strings similar to the existing implementation for example "metadata.name". I am just keeping them (as "resource.limits.cpu" ) to be consistent with the current implementation. So as I said they are fixed string, so could be anything "resource.limits.cpu" or "cpu" but we need to decide on one. 

Regarding implementation details, I dont think any new struct is really required as the same thing can be achieved with existing structs. See implementation of no selectors approach: https://github.com/kubernetes/kubernetes/pull/24179

aveshagarwal replied:
@pmorie That said, I am open to introduce a new struct as you described if the existing ones can't be reused or there are any implementation/API issues I am not able to think.

pmorie replied:
@aveshagarwal We want to avoid creating new magic keys, so you can either just add a new option to pick from, or rewrite the third option.  Maybe it would be better to add a new one.

aveshagarwal replied:
@pmorie I dont understand, The only difference is whether to use "resource.limits.cpu" or this "cpu" from users perspective.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+| container_name/cpu_request | resources.requests.cpu |
+|container_name/memory_request | resources.requests.memory |
+
+Since environment variables are container scoped, so container name must
```
s/so/the/

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+
+| Path | Selectors |
+| -------------------- | -------------------|
+| container_name/cpu_limit | resources.limits.cpu |
```
It's not clear to me what 'path' is supposed to be here

aveshagarwal replied:
Lets say container name is: test-container and a user wants cpu limit exposed as a volume for this container. User can create a pod using:

```
apiVersion: v1
kind: Pod
metadata:
  name: kubernetes-downwardapi-volume-example
spec:
  containers:
    - name: test-container
      image: whatever 
      command: whatever
     resources:
        requests:
          memory: "64Mi"
          cpu: "250m"
        limits:
          memory: "128Mi"
          cpu: "500m"
      volumeMounts:
        - name: podinfo
          mountPath: /etc
          readOnly: false
  volumes:
    - name: podinfo
      downwardAPI:
        items:
          - path: "test-container/cpu_limit"
            containerFieldRef:
              fieldPath: resources.limits.cpu
```

Once the pod is created: the cpu limit will be exposed as `/etc/test-container/cpu_limit` inside the container.

pmorie replied:
@aveshagarwal In your example above, there's nothing that orients the `containerFieldRef` as applying to `test-container`.

aveshagarwal replied:
@pmorie In item, there could be blocks of path and containerFieldRef . For example:

```
       items:
          - path: "containerA/cpu_limit"
            containerFieldRef:
              fieldPath: resources.limits.cpu
          - path: "containerB/cpu_limit"
            containerFieldRef:
              fieldPath: resources.limits.cpu
```
So each containerFieldRef belongs to its path above it. Its no different than the relation between existing path and FieldRef.

pmorie replied:
@aveshagarwal That's not how the API works.  `path` is 'where does this thing go within the volume', the reference associated w/ the path should fully identify _what_ goes at the path.

aveshagarwal replied:
@pmorie it does. thats why the resulting volume path is /etc/test-container/cpu_limit

ncdc replied:
@aveshagarwal the point @pmorie is trying to make is that `path` is up to the user. If they want to mount the cpu limit at `/foo`, they must be allowed to do so.

aveshagarwal replied:
@ncdc I had discussion with pmorie, and his main point is that we should not parse "path". I am ok with that.

aveshagarwal replied:
Although my point is that a path with container name in it for container resources like /etc/test-container/foo looks more obvious and could be less confusing when exposing limits from multiple containers like /test-container1/foo and /test-container2/foo instead of /foo1 and /foo2. Yes it looses the flexibility that it can not be just /foo. But it can be /test-container/foo.

aveshagarwal replied:
But I agree with pmorie too that its good to have paths as opaque as possible not need to parse them.

ncdc replied:
The tenet here is that the user gets to choose where the volume is mounted. This can't be violated.

aveshagarwal replied:
Nobody is taking the flexibility away where is the volume mounted? Where is that implied here?

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+bytes, and cpu request or limit of `250m` (milicores) will be output as
+`256` of cpu shares.
+
+## Validations
```
Move the different validations to live with the description of their respective approaches

lavalamp commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,530 @@
+is composed of json tags similar to json paths used with kubectl
+([json](http://kubernetes.io/docs/user-guide/jsonpath/)). This proposal
+uses kubernetes' json path library to process the selectors to retrieve
+the values. As kubelet operates on internal objects (without json tags),
```
kubelet should start using external versions and not internal versions.

pmorie replied:
@lavalamp is that likely to happen before 1.3?  That seems like it would be a pretty huge refactor.

lavalamp replied:
@pmorie, @caesarxuchao is flipping components over as he has time. But
starting with controllers, so I wouldn't expect it for 1.3. OTOH, if it's
blocking something, maybe we can expedite.

On Tue, Apr 26, 2016 at 8:03 AM, Paul Morie <notifications@github.com>
wrote:

> In docs/design/downward_api_resources_limits_requests.md
> <https://github.com/kubernetes/kubernetes/pull/24051#discussion_r61102455>
> :
>
> > +to their use.
> > +
> > +#### JSONpath selectors
> > +
> > +Versioned objects in kubernetes have json tags as part of their golang fields. Although currently internal objects
> > +in kubernetes also have json tags but these tags should be removed in
> > +future (see [3933](https://github.com/kubernetes/kubernetes/issues/3933)
> > +for discussion). So for discussion in this proposal, we assume that
> > +internal objects do not have json tags. In the first two approaches
> > +(full and partial json selectors), when a user creates a pod and its
> > +containers, the user specifies a json path selector in the pod's
> > +spec to retrieve values of its limits and requests. The selector
> > +is composed of json tags similar to json paths used with kubectl
> > +([json](http://kubernetes.io/docs/user-guide/jsonpath/)). This proposal
> > +uses kubernetes' json path library to process the selectors to retrieve
> > +the values. As kubelet operates on internal objects (without json tags),
>
> @lavalamp <https://github.com/lavalamp> is that likely to happen before
> 1.3? That seems like it would be a pretty huge refactor.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24051/files/54f11d6e55a30483482e968d7cdfda47ab1d0908#r61102455>
>

aveshagarwal replied:
@lavalamp, for this work (targeted for 1.3), if we use json path selectors for accessing resource values in kubelet, we would need to convert internal objects to versioned objects unless kubelet starts using external versions. If we are ok with that then not a blocker. Also not a blocker, if we just go with no selectors (or magic keys) approach.

lavalamp replied:
No arguments.

aveshagarwal commented:
Pushed an updated.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+3. In this approach, users specify fixed strings to retrieve
+resources limits and requests and do not specify any json path
+selectors. This approach is similar to the existing downward API
+implementation approach. The advantages of this approach is that it is
```
s/is/are/

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+#### JSONpath selectors
+
+Versioned objects in kubernetes have json tags as part of their golang fields.
+Although currently internal objects in kubernetes also have json tags but these
```
I would rewrite this as 'Currently, objects in the internal API have json tags, but it is planned that these will eventually be removed`

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+containers, the user specifies a json path selector in the pod's
+spec to retrieve values of its limits and requests. The selector
+is composed of json tags similar to json paths used with kubectl
+([json](http://kubernetes.io/docs/user-guide/jsonpath/)). This proposal
```
Does the proposal really use the library, or is it that some of the approaches discussed could use the library?

aveshagarwal replied:
Full and partial selector approaches use json path library.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+of the versioned object, so it seems more work from the first solution
+unless there is another way without requiring the versioned object.
+
+So there is a one time conversion cost associated with the first (full
```
I wouldn't call it one time, exactly.  It's once per API version conversion.

aveshagarwal replied:
@pmorie, I meant one time as in between a pod is created and terminated. Could you elaborate on "once per API version conversion"?

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+This table shows how selectors can be used for various requests and
+limits to be exposed as environment variables.
+
+| Name | Selector |
```
I would call the first column `information type` or something, and change the values for this column to `CPU limit` from `CPU_LIMIT`, etc.

aveshagarwal replied:
These are environment variable (I think I forgot to fix it here as in other places I am using env var name)and path names as I wanted to have something that I can use in examples below.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+This table shows how selectors can be used for various requests and
+limits to be exposed as volumes.
+
+| Path | Selector |
```
Same thing here - this table is really just describing how to address different information with a selector rather than dictating a path.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
````diff
@@ -0,0 +1,537 @@
+```
+
+#### Examples
+
````
Add a note about what each example is.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+
+| Name | Resource |
+| -------------------- | -------------------|
+| CPU_LIMIT | cpu_limit |
```
I would rather use a syntax for magic keys that would also be a valid selector in the future, basically the same selector as for other approaches, just implemented as a magic key.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+Also verify that container name is provided with volumes.
+
+## Pod and container resource access
+Pod resources will always be accessed with `type ObjectFieldSelector` object in
```
I'm not sure I understand what 'Pod resources' refers to in this sentence.

aveshagarwal replied:
Like metdata.name, status.IP or any existing (not part of downward API yet) or new pod level resources in the future.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,537 @@
+cgroups output format, i.e. cpu in cpu shares (cores multiplied by 1024
+and rounded to integer) and memory in bytes. For example, memory request
+or limit of `64Mi` in the container spec will be output as `67108864`
+bytes, and cpu request or limit of `250m` (milicores) will be output as
```
millicores

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,542 @@
+
+Partial json path selectors specify paths to resources limits and requests
+relative to the container spec. These will be implemented by introducing a
+`ContainerFieldSelector` (json: `containerFieldRef`) to extend the current
```
Probably we should specify that it's the `containerSpecFieldRef`. Just in case there's a case for adding status in the future.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,542 @@
+
+| Env Var Name | Resource |
+| -------------------- | -------------------|
+| CPU_LIMIT | resources.limits.cpu |
```
Given that `resource` is in the field name, I wouldn't expect it to also be in the values.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,542 @@
+| -------------------- | -------------------|-------------------|
+| Full selectors | `ObjectFieldSelector` | `ObjectFieldSelector`|
+| Partial selectors | `ObjectFieldSelector`| `ContainerFieldRef` |
+| No selectors | `ObjectFieldSelector`| `ResourceFieldRef` |
```
This still has selectors -- at least limits vs. requests.

aveshagarwal replied:
I think I can change it to magic keys approach instead of saying no selectors if that sounds better.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,542 @@
+     // namespace are supported.
+     FieldRef *ObjectFieldSelector `json:"fieldRef, omitempty"`
+     // Selects a field of the container: only resources limits and requests
+     // (cpu, memory) are currently supported.
```
Should specify the correct syntax in the documentation comment (resources.limits.cpu, etc.).

bgrant0607 commented:
I'd like to wrap this up. The clock is winding down on 1.3.

I'm in favor of `containerSpecFieldRef`. It eliminates the complicated part of the selector (choosing the container) without making it overly specific (just resources). That has the disadvantage of making the functionality somewhat less discoverable, assuming not all fields would be supported initially, but we really, really, really need to improve our help, documentation, examples, etc. Besides, even the most specific proposed form (resourceRef) doesn't help this problem beyond what `containerSpecFieldRef` provides.

Any counterarguments? @pmorie @lavalamp

aveshagarwal commented:
@bgrant0607 +1 from me, I also prefer containerSpecFieldRef. Lets see what others have to say. Once I get feedback from others, I will update it.

pmorie commented:
@bgrant0607 @aveshagarwal I am fine with that.  @aveshagarwal Could you make sure that all the existing feedback is addressed so @bgrant0607 can do the final review?

aveshagarwal commented:
Addressed comments and updated this PR.

aveshagarwal commented:
Updated. @bgrant0607 PTAL.

aveshagarwal commented:
Ran update*.sh and it should  fix tests.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,552 @@
+type EnvVarSource struct {
+   // Required: Selects a field of the container: only resources limits and
+   // requests (cpu, memory) are supported.
+   ContainerFieldRef *ContainerFieldSelector `json:"containerSpecFieldRef,omitempty"`
```
Also update the Go field name. The type name could probably stay the same.

aveshagarwal replied:
I agree and I have changed both golang field name and type name and will be in the next update. thanks.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,552 @@
+// Only one of its fields may be set.
+type EnvVarSource struct {
+   // Required: Selects a field of the container: only resources limits and
+   // requests (cpu, memory) are supported.
```
Please document the exact expected syntax in the comment.

Also, this field can't be required.

aveshagarwal replied:
Agree.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,552 @@
+of `limits.cpu`, `limits.memory`, `requests.cpu` and `requests.memory`.
+Also verify that container name is provided with volumes.
+
+## Pod and container level resource access
```
English is right-associative. Pod-level and container-level should be hyphenated when used as compound adjectives.

aveshagarwal replied:
ok

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,552 @@
+## Output Format
+
+The output format for resources limits and requests will be same as
+cgroups output format, i.e. cpu in cpu shares (cores multiplied by 1024
```
I didn't notice this before. What is the rationale for using the cgroup format? It's not going to match the actual number of shares in the case of, for instance, BestEffort containers. 
cc @vishh

bgrant0607 replied:
It also seems like shares will be harder to use, such as to set thread-pool sizes.

aveshagarwal replied:
I think the main reason to have similar format as cgroup was to provide similar experience (in terms of format) for applications if they have been using cgroup limits available inside containers.

@derekwaynecarr Would you like to add something here?

aveshagarwal replied:
As per this document: http://kubernetes.io/docs/user-guide/compute-resources/

```
The spec.container[].resources.limits.cpu is multiplied by 1024, converted to an integer, and used as the value of the --cpu-shares flag to the docker run command.
The spec.container[].resources.limits.memory is converted to an integer, and used as the value of the --memory flag to the docker run command.
```

So the value of cpu-share and memory (passed to docker) is output as memory and cpu limits. The resource requests are also computed similar ways and output in the similar format.

bgrant0607 replied:
I do not expect that document to remain true as resource QoS evolves:
https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/resource-qos.md

For instance, we know that the cgroup hierarchy will get more complex.

Shares are an implementation detail. Most legacy applications, which run outside containers as well, don't look at shares, and I think it's a bad idea for them to start looking at it. 

Managing CPU is complex: QoS, normalization across heterogeneous architectures, shares, quota, cpusets, core masks, nice, scheduling policies, ... 

What applications need is an abstraction that tells them:
* how much computing power is available, by some useful measure
* how many physical threads are available (e.g., to decide whether spinlocks will work, the min. number of OS threads needed)

bgrant0607 commented:
@justinsb Are you interested in taking a look?

bgrant0607 commented:
No real use cases, such as GOMAXPROCS or -Xmx?

aveshagarwal commented:
I have updated the doc with an example showing how to use memory limits with Xmx. I am not sure GOMAXPROCS can be assigned the value of any cpu limits or requests as it is number of cpus/cores, whereas limits/requests.cpu are a share.  Also updated this proposal based on other comments.

aveshagarwal commented:
@bgrant0607 @pmorie PTAL

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,577 @@
+
+As an application author, I want to be able to use cpu or memory requests and
+limits to configure the operational requirements of my applications inside containers.
+For example, Java based applications can configure and adjust their heap size.
```
Please add a little more detail -- something like: "Java applications expect to be made aware of the available heap size via a command line argument to the JVM.  Example: `java -Xmx <heap size>`"

bgrant0607 commented:
Re. GOMAXPROCS: Exactly my point. If we don't have a value that can be used to set that, we're doing something wrong.

aveshagarwal commented:
@bgrant As the goal in this proposal is to expose resources.limits.cpu, my understanding is (and I also discussed with @derekwaynecarr ) that resources.limits.cpu, in its current form, might not be the value that could be assigned to GOMAXPROCS, unless I am misunderstanding grossly. If there is something else in kube that is more aligned with GOMAXPROCS and needs to be exposed via downward API, may be I am not aware.

aveshagarwal commented:
I am sorry I meant @bgrant0607.

derekwaynecarr commented:
@bgrant0607 - RE: `GOMAXPROCS` in a Kubernetes cluster environment.  

Post Go 1.5+, `GOMAXPROCS` defaults to the value based on the return value of `runtime.NumCPU()`.  As a result, the default value that a Go program will see is dependent upon the node it was scheduled independent of the CPU request/limit specified in the pod definition.  So for example, if a container requested 2.5 cores, and it landed on a 8 core box, `GOMAXPROCS` will default to 8, and it's not actually clear to me reducing `GOMAXPROCS` is desired or not.  We do not peg to particular cpu-sets, and CPU shares are evaluated across all cores.  The interesting thing about `GOMAXPROCS` is that the value is expressed as an int, and I imagine a number of applications would want to get CPUs in integer precision or in millicores.

@aveshagarwal : RE the output format

I was looking at how Windows is implementing docker run --cpu-shares and came across this:
https://msdn.microsoft.com/en-us/virtualization/windowscontainers/management/manage_resources

I was surprised that default shares per process is 5000, which makes the normal linux mapping of 1024 annoying and potentially error-prone in a multi-runtime world.

I think I would prefer a model where the user could add additional arguments to a `ContainerRefFieldSelector` for values that select a `resource.Quantity`.  For example, I think the user should be able to specify the unit of measurement to output the value in and a quantization that should be used.

So for example, in the GOMAXPROCS example, I want CPU in cores with a quantization `RoundHalfUp`.

```
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh","-c", "env" ]
      resources:
        requests:
          memory: "64Mi"
          cpu: "250m"
        limits:
          memory: "128Mi"
          cpu: "500m"
      env:
        - name: CPU_LIMIT
          valueFrom:
            containerSpecFieldRef:
              fieldPath: resources.limits.cpu
              units: "cores"
              quantization: "RoundHalfUp"
```

So in this case, the units would be in cores. the quantization would say to round up, so the value for CPU_LIMIT would go to 1 in this case.


So for CPU, I would support units: "millicores", "cores"
For memory, I would support units: "bytes"

For quantization strategies, I would want to support "RoundHalfUp", "RoundHalfDown", "None" initially.

Thoughts?

pmorie commented:
What I'm still missing about this is whether there is a way to determine number of hardware threads to correctly set GOMAXPROCS from cpu shares, period.  is it number of 'whole' cores?

derekwaynecarr commented:
@pmorie - it must be in terms of whole cores.  I think if we had cpu-sets we would be better able to support setting GOMAXPROCS accurately, but I think the quantization and units approach is probably a better long term approach to support applications that want values in fixed integer ranges or something more fractional.

ncdc commented:
@derekwaynecarr's suggestion about quantization and rounding seems reasonable to me.

aveshagarwal commented:
@derekwaynecarr Sounds good to me.

derekwaynecarr commented:
I am not a math library expert by any means, so naming can be tweaked of course.

aveshagarwal commented:
@derekwaynecarr I just had a thought, is it possible that rounding the exposed values might cause inconsistencies with what was originally limited or requested for applications?

aveshagarwal commented:
@derekwaynecarr Or IOW, that allowing users to rounding might be error prone for applications?

pmorie commented:
I don't think rounding is problematic as long as it yields a value _under_ the actual request / limit.

derekwaynecarr commented:
To do GOMAXPROCS scenario, you need to round up.. If limit is500m,you want 1 for GOMAXPROCS not .5.

pmorie commented:
@derekwaynecarr ugh, you're right... in that case.  What about this one -- if the limit is 4500m, what do you want to see as a user? 4 cores or 5?

derekwaynecarr commented:
Whatever makes my application perform better which I am unable to answer
without knowing the application.  Go is hard to get right here because
increasing max procs can make an application perform worse depending on the
application.  Either way, I think this gives the right expressability and
is portable across runtimes.

On Monday, May 16, 2016, Paul Morie <notifications@github.com> wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> ugh, you're right...
> in that case. What about this one -- if the limit is 4500m, what do you
> want to see as a user? 4 cores or 5?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24051#issuecomment-219520623>
>

aveshagarwal commented:
@bgrant0607 Any feedback on @derekwaynecarr's suggestion above?

bgrant0607 commented:
@pmorie @derekwaynecarr GOMAXPROCS was an example. Any application that uses some kind of thread pool is going to need to decide how many threads to create, and on a shared machine using the physical/virtual number of hardware threads is probably not the best choice. Of course, the exact choice is up to the application and/or app operator, but it would be easier to specify by env var, flag, etc. if we provided cores rather than pseudo-shares. 

It's true we don't manage cpusets, cpu masks, etc. today, but I expect we will eventually manage core allocation implicitly, based on QoS class.

What this discussion does make clear is that legacy applications would benefit from resource-specific functionality. Therefore, I think we should go with resourceRef, supporting {limits,requests}.{cpu,memory}.

We don't need to add all the features in the initial version, so long as we have a way to add them later.

derekwaynecarr commented:
@bgrant0607 - were you good with the units and rounding options for each
compute resource?  I think that should allow the right flexibility instead
of shares.

On Tuesday, May 17, 2016, Brian Grant <notifications@github.com> wrote:

> @pmorie <https://github.com/pmorie> @derekwaynecarr
> <https://github.com/derekwaynecarr> GOMAXPROCS was an example. Any
> application that uses some kind of thread pool is going to need to decide
> how many threads to create, and on a shared machine using the
> physical/virtual number of hardware threads is probably not the best
> choice. Of course, the exact choice is up to the application and/or app
> operator, but it would be easier to specify by env var, flag, etc. if we
> provided cores rather than pseudo-shares.
>
> It's true we don't manage cpusets, cpu masks, etc. today, but I expect we
> will eventually manage core allocation implicitly, based on QoS class.
>
> What this discussion does make clear is that legacy applications would
> benefit from resource-specific functionality. Therefore, I think we should
> go with resourceRef, supporting {limits,requests}.{cpu,memory}.
>
> We don't need to add all the features in the initial version, so long as
> we have a way to add them later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24051#issuecomment-219823446>
>

bgrant0607 commented:
@derekwaynecarr 

What does RoundHalfUp mean? Round off? I'd probably start with that and make it the default when we add more options later: Ceiling, Floor, etc. Rounding probably needs to be different for values that would round to zero, but we could punt on that initially.

As far as units go, I'd like to keep them resource-independent, like Quantity, so just `scale`, not `units`. I'd use the same suffixes supported by pkg/api/resource/quantity.go: 1, 1m, 1Mi, 1e6, etc.

aveshagarwal commented:
I am summarizing here before I update this proposal to make sure I get it right. Please let me know if am stating anything incorrectly.

So we are going with:

```
ResourceFieldRef *ResourceFieldSelector `json:"resourceFieldRef,omitempty"`
```

And it would look like:

```
type ResourceFieldSelector struct {
        Name string `json:"containerName,omitempty"`
        Resource string `json:"resource"`
       Quantization string  `json:"quatization,omitempty"`
       Scale string `json:"scale,omitempty"`

}
```

**Quatization**: Right now it will have only RoundHalfUp (or Roundoff I am ok with either). It means for cpu that 3300 milicore will become 4000  or in cores 4. By default, currently the cpu value stays same without any rounding. So when a user explicitly configures quantization as Roundoff, then only it will take affect. 

Another option is to make Roundoff default now too.

Also my understanding is that right now it only applies to cpu not memory. 

Future work:  add Ceiling, Floor, and at that time, Roundoff would become the default.

**Scale**: 1, 1m, 1Mi, 1e6 as per pkg/api/resource/quantity.go, by default  1.

Let me know if I am missing something.

@bgrant0607 @derekwaynecarr PTAL.

pmorie commented:
@aveshagarwal @derekwaynecarr why `RoundHalfUp` instead of just `RoundUp`, afaict this strategy is just grade-school rounding up.

aveshagarwal commented:
@pmorie RoundUp sounds good.

pmorie commented:
@aveshagarwal I would also suggest `ContainerName` instead of `Name`

aveshagarwal commented:
@pmorie sounds good.

aveshagarwal commented:
In pkg/api/resource/quantity.go, scale constants are:

```
const (
        Nano  Scale = -9
        Micro Scale = -6
        Milli Scale = -3
        Kilo  Scale = 3
        Mega  Scale = 6
        Giga  Scale = 9
        Tera  Scale = 12
        Peta  Scale = 15
        Exa   Scale = 18
)
```

So I am thinking to propose scale values as: nano, micro, mili, kilo, mega, giga, tera, peta, and exa  and the output will be suffixed accordingly like m, Mi.

pmorie commented:
@aveshagarwal why not use the constants already in the API?

aveshagarwal commented:
@pmorie could you elaborate what do you mean by that?

pmorie commented:
@aveshagarwal If we already have constants for these magnitudes, why not use those constants, instead of making new string versions for the same thing?

aveshagarwal commented:
@pmorie are you talking about unit suffixes like m, Mi, Ki, k?

pmorie commented:
I'm talking about https://github.com/kubernetes/kubernetes/pull/24051#issuecomment-220142018

aveshagarwal commented:
@pmorie  you said "If we already have constants for these magnitudes"? which constants you are referring to?

pmorie commented:
@aveshagarwal I'm talking about the constants you referenced in the comment iI linked.  What am I missing?

aveshagarwal commented:
@pmorie, could you tell explicitly what I changed then?

pmorie commented:
@aveshagarwal it appeard to me you were proposing that we make new constants typedef'd to string in addition to the numerical constants in api/resource.  If that is accurate, I am asking what the point of doing that is, and why we shouldn't just reuse the numberical constants.  Here's the specific comment I'm referring to:

> In pkg/api/resource/quantity.go, scale constants are:

https://github.com/kubernetes/kubernetes/pull/24051#issuecomment-220142018

aveshagarwal commented:
@pmorie, As Scale is string, I was saying what user input should be for example, nano/Nano for Nano. Is something wrong with it?

pmorie commented:
@aveshagarwal I am asking why scale has to be string, since it is an int elsewhere.  It might be arguable that strings and descriptive constant values would be a better user experience, but that's your argument to make :)

cc @derekwaynecarr

aveshagarwal commented:
@pmorie I am not very stringent about it. But I thought entering nano/Nano seems more user readable than -9 or 3 (which is actually 10^-9 or 10^3). May be its just me.

pmorie commented:
@aveshagarwal I **think** this is the first place that scale is actually exposed as such in the API, so it could go either way, precedent-wise.

derekwaynecarr commented:
@bgrant0607 @aveshagarwal @pmorie `RoundUp` is fine.  The general requirement is I need to be able to specify rounding rules to support things like the max procs discussion.  For scale, we just need a means to denote the scale exponent.  I suspect a lot of people are not going to know the nuance of -3, 0, 3, etc. and they would appreciate a token of some kind.  I assume an empty token would default to scale: 0?  Also, we should clarify rounding would happen on the scaled value.

derekwaynecarr commented:
@bgrant0607 @pmorie @aveshagarwal Actually, if we just call `ScaledValue` it defaults to ceiling, so that would work fine for CPU, and maybe we can punt on the rounding stuff for now altogether?  So I think if we just support a scale field that is enough.  So 250m if requested with scale 0 will just get 1.  That's probably simplest.

pmorie commented:
If we expose scale directly I think human readable is best; I was initially under the impression that it was exposed in numeric form and didn't want to have two different APIs for the same thing.  Apologies for the confusion.

bgrant0607 commented:
@derekwaynecarr @pmorie @aveshagarwal I'm fine with ceiling as the default and only rounding policy for now. I agree we shouldn't even add a field for that yet -- that's what I intended.

bgrant0607 commented:
@derekwaynecarr @pmorie @aveshagarwal Regarding scale: How about we call it `divisor` and use resource Quantity to express it, defaulting to "1"?

derekwaynecarr commented:
My math-fu is not all there right now.  I assume the divisor is independent
of the internal scale factor.  For CPU I will either want core (1) or
millicore (3)  - assuming it's not relative to the internal scale factor.

For memory, is it not more awkward if I want input as Mi instead of bytes?
Having a negative divisor is awkward and makes me think I will get a
negative result as a naive user.

Personally I think scale makes more sense.

On Wednesday, May 18, 2016, Brian Grant <notifications@github.com> wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> @pmorie
> <https://github.com/pmorie> @aveshagarwal
> <https://github.com/aveshagarwal> Regarding scale: How about we call it
> divisor and use resource Quantity to express it, defaulting to "1"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24051#issuecomment-220189475>
>

pmorie commented:
I would rather explain `exponent`, or `scale` to users than `divisor`.

derekwaynecarr commented:
Unless you thought you would give a divisor that was always non negative
and you would have to do .25 or something.... Think scale is still simpler.

On Wednesday, May 18, 2016, Derek Carr <decarr@redhat.com> wrote:

> My math-fu is not all there right now.  I assume the divisor is
> independent of the internal scale factor.  For CPU I will either want core
> (1) or millicore (3)  - assuming it's not relative to the internal scale
> factor.
>
> For memory, is it not more awkward if I want input as Mi instead of
> bytes?  Having a negative divisor is awkward and makes me think I will get
> a negative result as a naive user.
>
> Personally I think scale makes more sense.
>
> On Wednesday, May 18, 2016, Brian Grant <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> @derekwaynecarr <https://github.com/derekwaynecarr> @pmorie
>> <https://github.com/pmorie> @aveshagarwal
>> <https://github.com/aveshagarwal> Regarding scale: How about we call it
>> divisor and use resource Quantity to express it, defaulting to "1"?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/24051#issuecomment-220189475>
>>
>

aveshagarwal commented:
I agree with @derekwaynecarr that scale seems more appropriate. And for cpu, mostly 0 (cores) and 3 (milicores) can serve the purpose. Also I agree with that ScaledValue does ceiling by default so should do the job.

resource.Quantity can be expressed in 3 ways: DecimalSI, BinarySI, DecimalExponent and I am not sure for divisor we would want that flexibility. Also there is no direct divide operation in pkg/api/resource that operates on 2 resource.Quantities. I infact wonder what is the advantage of using divisor as resource.Quantity?

aveshagarwal commented:
Just to add the way `scale` works currently in pkg/api/resource, its like converting from one unit to another and this is something we want for cpu (millicore to cores as needed) as I understand.

pmorie commented:
scale WFM

bgrant0607 commented:
@derekwaynecarr @pmorie @aveshagarwal 

My concern with "scale" is that that term is used in another context (i.e., scaling, autoscaling, scale subresource).

Quantity supports arithmetic, no?

Let's say my memory is 100Mi and I specify a divisor of 1Mi: ceil(100Mi / 1Mi) = 100.

Let's say my cpu is 1200m and I specify a divisor of 1: ceil(1200m / 1) = 2

Let's say my cpu is 500m and I specify a divisor of 1m: ceil(500m / 1m) = 500

The result would always be an integer.

bgrant0607 commented:
Don't all Quantity values already have to be nonnegative?

aveshagarwal commented:
@bgrant0607 As I understand that I dont think if a,b are 2 Quantities, direct a/b is possible, but yes  a.ScaledValue(0)/b.ScaledValue(0) is ofcourse possible because ScaledValue converts them to int64.

bgrant0607 commented:
Div could be implemented similarly to Add, Sub, and Neg:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/resource/quantity.go#L345

aveshagarwal commented:
@bgrant0607 Right, I saw that I just meant that div is not there currently.

pmorie commented:
@bgrant0607 how does `exponent` strike you?

bgrant0607 commented:
@pmorie Let's say I specified memory request of "100Mi". What exponent would I specify to get "100" as the result?

bgrant0607 commented:
@pmorie An even more concrete example: I want to specify `-Xmx$(HEAP_SIZE)m`. How would I do that?

aveshagarwal commented:
IMHO, cpu and memory are quantities with units, why would we want to expose them unit less? Also isn't the job of application to manipulate them whatever way they want to use them?

derekwaynecarr commented:
@bgrant0607 - internally a quantity can be negative.  externally, we validate against negatives.  I think `divisor` makes sense as described.

Heap size is specified in `DecimalSI` and not `BinarySI` in the JVM if I recall.  As long as the format is consistent with the `divisor`, I think its okay.  I would want @bparees to weigh in.

You would specify heap size as follows:

```yaml
spec:
  containers:
    - name: test-container
      resources:
        requests:
          memory: "64M"
          cpu: "250m"
        limits:
          memory: "128M"
          cpu: "500m"
      env:
        - name: MIN_HEAP_SIZE
          valueFrom:
            containerSpecFieldRef:
              fieldPath: resources.requests.memory
              divisor: 1M
        - name: MAX_HEAP_SIZE
          valueFrom:
            containerSpecFieldRef:
              fieldPath: resources.limits.memory
              divisor: 1M
```

Assuming we include units, the command would declare:

` -Xms${MIN_HEAP_SIZE} -Xmx${MAX_HEAP_SIZE}`

If we exclude units, the command would declare the unit ourself (m or M is valid I think):

` -Xms${MIN_HEAP_SIZE}m -Xmx${MAX_HEAP_SIZE}m`

@aveshagarwal - we have some code in openshift that does multiply/divide here: https://github.com/openshift/origin/blob/master/pkg/quota/admission/clusterresourceoverride/admission.go#L210  ; I would be a fan of getting it in `quantity.go`.

derekwaynecarr commented:
Hit enter too soon on previous comment, I think I would want to exclude units by default.

aveshagarwal commented:
@derekwaynecarr Yeah if we do not want to include altogether that's fine. Basically this means that the precision will always be based on the divisor.

aveshagarwal commented:
So I will update this proposal with:

type ResourceFieldSelector struct {
        ContainerName string `json:"containerName,omitempty"`
        Resource string `json:"resource"`
        Divisor resource.Quantity `json:"divisor,omitempty"`

}

By default divisor will be 1 (mean bytes). And there will be no units displayed with the exposed value of cpu and memory.

aveshagarwal commented:
And for cpu, the default value 1 will mean cores.

aveshagarwal commented:
@bgrant0607 @derekwaynecarr @pmorie PTAL
I have pushed updated proposal and it includes support for divisor with ResourceFieldSelector.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,644 @@
+| cpu_request | requests.cpu |
+| memory_request | requests.memory |
+
+Since environment variables are container scoped, the container name must
```
The container name must be specified because volume are pod scoped rather than container scoped.

aveshagarwal replied:
yeah will be fixed in next update soon.

aveshagarwal commented:
fixed tests.

pmorie commented on docs/design/downward_api_resources_limits_requests.md:
````diff
@@ -0,0 +1,646 @@
+```
+
+Note that the value of divisor by default is `1`. Now inside the container,
+the HEAP_SIZE (in bytes) and GOMAXPROCS (in cores) could be exported as:
````
Can we see heap size in Megs in a different example (or in a volume in this pod) ?

pmorie commented:
@aveshagarwal The information here is good.  My suggestions are purely additive:

There needs to be a passage stating which approach we will implement and why.  It would be fine with me to add it at the end, and keep from spending time discussing how to mutate the text of this proposal so we can update the PR.

I **think** other than that, this LGTM.  Thanks a lot for the proposal.

aveshagarwal commented:
@pmorie updated PR.

pmorie commented:
@aveshagarwal LGTM

@bgrant0607 PTAL

bparees commented:
>  As long as the format is consistent with the divisor, I think its okay. I would want @bparees to weigh in.

agree and also agree that we don't want units by default, i think there are lots of cases where the unit would make the value harder to use in a config.

pmorie commented:
I do not want units period initially.

aveshagarwal commented:
@bparees yeah there are no units in the output now in the proposal with the selected approach.

bgrant0607 commented on docs/design/downward_api_resources_limits_requests.md:
```diff
@@ -0,0 +1,651 @@
+| memory_request | requests.memory |
+
+Volumes are pod scoped, the container name must be specified as part of
+`containerSpecFieldRef` with them.
```
resourceFieldRef

bgrant0607 commented:
Thank you. One nit, but LGTM. Feel free to reapply lgtm if you fix the nit.

pmorie commented:
@aveshagarwal, send a follow-up PR to fix that nit and I will tag it.
Thanks again.

On Friday, May 20, 2016, k8s-merge-robot <notifications@github.com> wrote:

> Merged #24051 <https://github.com/kubernetes/kubernetes/pull/24051>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24051#event-666933813>
>

aveshagarwal commented:
@pmorie https://github.com/kubernetes/kubernetes/pull/25962
